config/policy.go

```diff
@@ -333,115 +350,190 @@ func NewPolicyFromProto(pb *configpb.Route) (*Policy, error) {
-		AllowPublicUnauthenticatedAccess: p.AllowPublicUnauthenticatedAccess,
-		AllowAnyAuthenticatedUser:        p.AllowAnyAuthenticatedUser,
-		Timeout:                          timeout,
-		IdleTimeout:                      idleTimeout,
-		AllowWebsockets:                  p.AllowWebsockets,
-		AllowSpdy:                        p.AllowSPDY,
-		TlsSkipVerify:                    p.TLSSkipVerify,
-		TlsServerName:                    p.TLSServerName,
-		TlsUpstreamServerName:            p.TLSUpstreamServerName,
-		TlsDownstreamServerName:          p.TLSDownstreamServerName,
-		TlsCustomCa:                      p.TLSCustomCA,
-		TlsCustomCaFile:                  p.TLSCustomCAFile,
-		TlsClientCert:                    p.TLSClientCert,
-		TlsClientKey:                     p.TLSClientKey,
-		TlsClientCertFile:                p.TLSClientCertFile,
-		TlsClientKeyFile:                 p.TLSClientKeyFile,
-		TlsDownstreamClientCa:            p.TLSDownstreamClientCA,
-		TlsDownstreamClientCaFile:        p.TLSDownstreamClientCAFile,
-		SetRequestHeaders:                p.SetRequestHeaders,
-		RemoveRequestHeaders:             p.RemoveRequestHeaders,
-		PreserveHostHeader:               p.PreserveHostHeader,
-		PassIdentityHeaders:              p.PassIdentityHeaders,
-		KubernetesServiceAccountToken:    p.KubernetesServiceAccountToken,
-		Policies:                         sps,
-		SetResponseHeaders:               p.SetResponseHeaders,
-	}
-	if p.IDPClientID != "" {
-		pb.IdpClientId = proto.String(p.IDPClientID)
-	}
-	if p.IDPClientSecret != "" {
-		pb.IdpClientSecret = proto.String(p.IDPClientSecret)
-	}
-	if p.Redirect != nil {
-		pb.Redirect = &configpb.RouteRedirect{
-			HttpsRedirect:  p.Redirect.HTTPSRedirect,
-			SchemeRedirect: p.Redirect.SchemeRedirect,
-			HostRedirect:   p.Redirect.HostRedirect,
-			PortRedirect:   p.Redirect.PortRedirect,
-			PathRedirect:   p.Redirect.PathRedirect,
-			PrefixRewrite:  p.Redirect.PrefixRewrite,
-			ResponseCode:   p.Redirect.ResponseCode,
-			StripQuery:     p.Redirect.StripQuery,
-		}
-	} else if p.Response != nil {
-		pb.Response = &configpb.RouteDirectResponse{
-			Status: uint32(p.Response.Status),
-			Body:   p.Response.Body,
-		}
-	} else {
-		to, weights, err := p.To.Flatten()
-		if err != nil {
-			return nil, err
-		}
-
-		pb.To = to
-		pb.LoadBalancingWeights = weights
-	}
-
+	clear(dest.RewriteResponseHeaders)
+	dest.RewriteResponseHeaders = slices.Grow(dest.RewriteResponseHeaders[:0], len(p.RewriteResponseHeaders))
 	for _, rwh := range p.RewriteResponseHeaders {
-		pb.RewriteResponseHeaders = append(pb.RewriteResponseHeaders, &configpb.RouteRewriteHeader{
+		dest.RewriteResponseHeaders = append(dest.RewriteResponseHeaders, &configpb.RouteRewriteHeader{
 			Header: rwh.Header,
 			Matcher: &configpb.RouteRewriteHeader_Prefix{
 				Prefix: rwh.Prefix,
 			},
 			Value: rwh.Value,
 		})
 	}
+}
+
+func copySrcToOptionalDest[T comparable](dst **T, src *T) {
```

Comment:
Can you help me understand this method? Why does it sometimes copy a pointer and sometimes copy a value?

Reply:
`src` is a pointer to a (non-pointer/non-optional) struct field, and `dst` is a pointer to a (pointer/optional) struct field. This does the following:
- If `*src` is a zero-valued `T`, `*dst` is set to nil. This is a best-effort preservation of optional semantics. fields for which a value of zero is valid will be unset anyway, but there is no way around this.
- Otherwise (`*src` has a non-zero value)
	- If `*dst` is a non-nil pointer, copy the value `*src` into `**dst`, to reuse the heap memory already pointed to by `*dst` (if possible; strings etc. may or may not end up needing to be resized)
	- If `*dst` is a nil pointer, set `*dst` to `src` directly. the pointer field in the destination struct now holds an address interior to the source struct.

Reply:
> * If `*dst` is a non-nil pointer, copy the value `*src` into `**dst`, to reuse the heap memory already pointed to by `*dst` (if possible; strings etc. may or may not end up needing to be resized)

This is the case I find confusing. In what circumstances is it better to copy the value rather than the pointer?

Or to put the question another way, if it's OK to alias parts of the source struct, why not always copy the `src` pointer?

Reply:
To reduce GC pressure. if we overwrite the original pointer, that memory needs to be garbage collected. If we copy the value, it doesn't. This allows using a sync.Pool to store commonly used fields that need to be heap allocated.

Reply:
OK, I think I understand — this trades additional memory usage for reduced GC time?

Reply:
Both, this ends up drastically reducing memory usage, which helps achieve the actual goal of reducing GC time.